### PR TITLE
fix logger type hint

### PR DESCRIPTION
**Pull request by tshu-w**
## What does this PR do?

fix logger type hint.

## Before submitting
- [ ] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [x] Did you make sure to update the documentation with your changes? (if necessary)
- [ ] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [ ] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)
- [x] Did you list all the breaking changes introduced by this pull request?

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: pytorch_lightning/trainer/connectors/logger_connector/logger_connector.py [hunk 1]
```diff
@@ -46,7 +46,7 @@ def __init__(self, trainer: 'pl.Trainer', log_gpu_memory: Optional[str] = None)
 
     def on_trainer_init(
         self,
-        logger: LightningLoggerBase,
+        logger: Union[bool, Iterable, LightningLoggerBase],
```
Comment (kandluis) [inline pytorch_lightning/trainer/connectors/logger_connector/logger_connector.py:49]: Is it possible to be more precise with the `Iterable` type here? Would `Iterable[LightningLoggerBase]` work?

**Review by kandluis — APPROVED** [review]

File: pytorch_lightning/trainer/connectors/logger_connector/logger_connector.py (hunk 1 above)
Comment (tshu-w) [reply pytorch_lightning/trainer/connectors/logger_connector/logger_connector.py]: @kandluis updated

**Review by carmocca — APPROVED** [review]

**Review by Borda — APPROVED** [review]
do we have enabled Mypy for this file?

**Comment by carmocca** [comment]
> do we have enabled Mypy for this file?

Yes: https://github.com/PyTorchLightning/pytorch-lightning/blob/master/setup.cfg#L170-L171

File: pytorch_lightning/trainer/connectors/logger_connector/logger_connector.py [hunk 2]
```diff
@@ -46,7 +46,7 @@ def __init__(self, trainer: 'pl.Trainer', log_gpu_memory: Optional[str] = None)
 
     def on_trainer_init(
         self,
-        logger: LightningLoggerBase,
+        logger: Union[bool, Iterable[LightningLoggerBase], LightningLoggerBase],
```
Comment (awaelchli) [inline pytorch_lightning/trainer/connectors/logger_connector/logger_connector.py:49]: don't we want to put the logger base before iterable?

**Review by awaelchli — APPROVED** [review]

File: pytorch_lightning/trainer/connectors/logger_connector/logger_connector.py (hunk 2 above)
Comment (carmocca) [reply pytorch_lightning/trainer/connectors/logger_connector/logger_connector.py:49]: Does this have an impact or just for looks? In any case, I agree

Comment (awaelchli) [reply pytorch_lightning/trainer/connectors/logger_connector/logger_connector.py:49]: for readability I guess, one could argument that ordering primitives first then composed types is easier to read.
just a nit

**Review by s-rog — APPROVED** [review]

File: pytorch_lightning/trainer/connectors/logger_connector/logger_connector.py (hunk 2 above)
Comment (kandluis) [reply pytorch_lightning/trainer/connectors/logger_connector/logger_connector.py:49]: Definitely a nit. I wonder if it's possible to configure the code formatter/linter to do this? Much easier to enforce for consistency, which would go a long way in aiding readability.

Comment (tshu-w) [reply pytorch_lightning/trainer/connectors/logger_connector/logger_connector.py]: @awaelchli updated.

**Comment by tshu-w** [comment]
Beyond this PR, I'm not sure how mypy works, but the type hint `'pl.Trainer'` for `trainer` should be updated too, there are many more issues like this, e.g. https://github.com/PyTorchLightning/pytorch-lightning/discussions/8285.
I'm using https://github.com/microsoft/pyright, and it will complain about it.
Screen Shot 2021-07-13 at 10 40 14

**Review by justusschock — APPROVED** [review]

**Comment by carmocca** [comment]
> Beyond this PR, I'm not sure how mypy works, but the type hint `'pl.Trainer'` for `trainer` should be updated too, there are many more issues like this, e.g. #8285.

Not familiar with `pyright`, but we use `mypy` and ignore the rule

https://github.com/PyTorchLightning/pytorch-lightning/blob/9d5ad7639cb0e694bb8634ece69235cfd330908d/pyproject.toml#L31-L32

As it doesn't understand properties set in connectors. Looks like the same issue (might be wrong)

**Pull request merged**